### ISSUE TITLE
[Caffe2] Enable Windows build with MKL-DNN

### ISSUE THREAD

**Issue by xkszltl** [pytorch/pytorch#15982]
## Bug

In MKLDNN cmake, MSVS build are disabled explicitly.
https://github.com/pytorch/pytorch/blob/12e0ed55b4327d41052651ddbae58506c5957ab8/cmake/Modules/FindMKLDNN.cmake#L16-L19

I think this is no longer the case as we can build mkldnn directly with Visual Studio 2017 and MKLDNN's official cmake doesn't complain about that
https://github.com/intel/mkl-dnn/blob/4bdffc2cb1c3d47df9604d35d2c7e5e47a13f1a6/cmake/OpenMP.cmake#L78-L81

But, after commenting out the `RETURN()`, the issue I'm encountering is about ideep.

```
pytorch\third_party\ideep\include\ideep\allocators.hpp(151): error C2039: 'posix_memalign': is not a member of '`global namespace''
pytorch\third_party\ideep\include\ideep\allocators.hpp(151): error C3861: 'posix_memalign': identifier not found
```

## Environment

 - PyTorch Version (e.g., 1.0): master
 - OS (e.g., Linux): Win10
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): cmake + ninja
 - Python version: 3.7.2
 - CUDA/cuDNN version: 10.0/7
 - GPU models and configuration: 1080Ti

**Comment by ezyang** [comment]
Should be simple enough to ifdef this into `_aligned_malloc` on Windows.

**Comment by peterjc123** [comment]
Just as @ezyang said, the build could be easily fixed. But my concern is that if openmp > 3 is not supported by MSVC, is it meaningful to enable this one? How can it run faster?  My suggestion is, if there is no additional openmp code required at the PyTorch side, what about we compile the ideep part using the compiler that support openmp > 3 (clang-cl, gcc and e.g.) and then link against the lib?

**Comment by xkszltl** [comment]
@peterjc123 Do they heavily rely on OpenMP 3? I do see things like:
https://github.com/intel/ideep/blob/c6eb7b911d81a72de9791838fcde113e0bfa3ec5/include/ideep/computations.hpp#L4233-L4237
which should provided still provided some benefits with degraded performance.
Also since the Win32 macro is there, is clang-cl still helpful?

Just ask, not digging deeply into the related code yet.

**Comment by xkszltl** [comment]
> Should be simple enough to ifdef this into `_aligned_malloc` on Windows.

Thanks!
Didn't expect that it can be done with brute force.
Will that you know if it's enough.

**Comment by peterjc123** [comment]
@xkszltl They mentioned that OpenMP 4.0 SIMD extensions are used and recommended their Intel C++ compiler. But I'm not sure if it's heavily used. https://github.com/intel/mkl-dnn#system-requirements As for the `WIN32` macro, we could try to change it to `#if defined(WIN32) && !defined(__clang__)` as long as it brings much speed improvement and little code modification.

**Comment by peterjc123** [comment]
@xkszltl I saw a speed difference between mkldnn on Linux and Windows(using MSVC) in this issue https://github.com/intel/mkl-dnn/issues/377. The speed difference is nearly 10x, which is not small. So I guess openmp > 3 should matter a lot.

**Comment by peterjc123** [comment]
@xkszltl Well, let's have a try first.

**Comment by xkszltl** [comment]
@ezyang I just realized we can define `WIN32` directly as the code is:
```
#if defined(WIN32)
      ptr = _aligned_malloc(size, alignment_);
#else
      int rc = ::posix_memalign(&ptr, alignment_, len);
      if (rc != 0)
        throw std::invalid_argument("Out of memory");
#endif
```

**Comment by xkszltl** [comment]
But `ideep/web.hpp` still asked for posix file `sched.h` which doesn't exist on Windows.

**Comment by peterjc123** [comment]
@xkszltl Could you please remove that header and see what's used there? If the usage is simple enough, we could port it to use std::thread.

**Comment by xkszltl** [comment]
It might not even be used.
I didn't see any related error after commenting of that `#include`

**Comment by xkszltl** [comment]
Unfortunately, ideep doesn't like MSVC

```
pytorch\third_party\ideep\include\ideep\utils.hpp(88): error C2059: syntax error: 'volatile'
pytorch\third_party\ideep\include\ideep\utils.hpp(192): error C3001: 'simd': expected an OpenMP directive name
pytorch\third_party\ideep\include\ideep/tensor.hpp(1379): error C3005: 'collapse': unexpected token encountered on OpenMP 'parallel for' directive

pytorch\third_party\ideep\include\ideep/tensor.hpp(1392): error C3005: 'collapse': unexpected token encountered on OpenMP 'parallel for' directive

pytorch\third_party\ideep\include\ideep/computations.hpp(5931): error C2039: 'time': is not a member of 'std'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\bitset(19): note: see declaration of 'std'
pytorch\third_party\ideep\include\TR_interface.h(17): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(17): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(17): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(19): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(19): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(19): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(19): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(21): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(21): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(21): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(21): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(22): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(22): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(22): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(22): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(24): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(24): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(24): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(24): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(26): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(26): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(26): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(26): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(29): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(29): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(29): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(29): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(31): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(31): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(31): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(31): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(32): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(32): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(32): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(32): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(33): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(33): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(33): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(33): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(34): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(34): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(34): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(34): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\TR_interface.h(36): error C3861: 'visibility': identifier not found
pytorch\third_party\ideep\include\TR_interface.h(36): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
pytorch\third_party\ideep\include\TR_interface.h(36): error C2374: '__attribute__': redefinition; multiple initialization
pytorch\third_party\ideep\include\TR_interface.h(17): note: see declaration of '__attribute__'
pytorch\third_party\ideep\include\TR_interface.h(36): error C2448: '__attribute__': function-style initializer appears to be a function definition

pytorch\third_party\ideep\include\ideep/distribute.hpp(56): error C3861: 'TR_available': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(62): error C3861: 'TR_init': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(68): error C3861: 'TR_init': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(73): error C3861: 'TR_get_world_size': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(78): error C3861: 'TR_get_rank': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(193): error C3861: 'TR_wait': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(207): error C3861: 'TR_test': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(211): error C3861: 'TR_test': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(216): error C3861: 'TR_set_urgent': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(221): error C3861: 'TR_barrier': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(226): error C3861: 'TR_finalize': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(256): error C3861: 'TR_allreduce': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(287): error C3861: 'TR_iallreduce': identifier not found
pytorch\third_party\ideep\include\ideep/distribute.hpp(311): error C3861: 'TR_bcast': identifier not found
```

**Comment by lilinxiong** [comment]
I had the same problem, By looking at the build log, find is **ideep** build failure, is not **mkldnn**  in the [intel/ideep#26](https://github.com/intel/ideep/issues/26) as if ideep doesn't support installation on windows. Does that mean in windows no use mkldnn?

**Comment by peterjc123** [comment]
cc @ezyang Are we using ideep in PyTorch or Caffe2?

**Comment by ezyang** [comment]
I'm not sure.

**Comment by xsacha** [comment]
MKL-DNN is enabled in the Linux prebuilts/nightlies but not in the Windows one.
Is it easy to just switch it on or is there bugs in the Windows implementation right now?

**Comment by xkszltl** [comment]
ideep does not compile in vs since vs only support openmp2

**Comment by peterjc123** [comment]
Yes, but if it is not an essential component, then we can bypass it.

**Comment by peterjc123** [comment]
Well, I searched for the keyword `ideep` in the project. Looks like it is used in both ATen and Caffe2. We have to fix the Windows build for ideep first. However, it is not an easy task.

**Comment by xkszltl** [comment]
Just wondering who is maintaining ideep? I thought it's Intel...

**Comment by peterjc123** [comment]
@xkszltl Maybe intel is maintaining it. Well, it isn't that important. Actually, I tried to fix the repo gradually by replacing `__attribute__((visibility("default")))` with `__dllspec(dllexport)` and hiding unsupported OpenMP clauses, but then I'm blocked by the usage of the inline assembly code in that project.

**Comment by neFAST** [comment]
So are we left with Windows cpu inference running one order of magnitude slower than on a Linux machine?

**Comment by peterjc123** [comment]
@neFAST Yes, it's part of the reason. But I  think there are far more factors that may affect this, such as the missing functionalies like openmp, blas (fbgemm) , and sometimes misconfiguration. Actually, I have little knowledge of how pytorch jit works so I don't know the exact reason of the speed difference between Windows and Linux.

**Comment by peterjc123** [comment]
I opened an issue at ideep: https://github.com/intel/ideep/issues/35. Hope that they will support ideep on Windows.

**Comment by shepherd1530** [comment]
@peterjc123 I saw the ideep issue has been resolved but I still can get MKL-DNN included with Pytorch binaries from both Anaconda and pip.

**Comment by peterjc123** [comment]
@shepherd1530 It is available in the 1.5.0 and the nightly packages.
